### DISCUSSION TITLE
LSP + Unsuitable Performance

### DISCUSSION THREAD

**Discussion by trev-dev** [joaotavora/eglot#993] [Q&A] [answered]
Hey all,

I can't be the only one who seems to struggle with LSP performance, regardless of the front-end client I'm using. People seem to be happy enough with things like Emacs & Eglot/LSP-Mode, so I must be dropping the ball by mostly sticking with the default settings.

I want to be able to do the following:

1. Move around the file without my point stuttering/freezing
2. Still be able to call up eldoc/help at point functions if I need them
3. Still see my code linted with flymake

Other features such as highlighting symbols and whatnot seem nice, but if they're contributing to my problem, I'm not interested in them.

I know I can just go read the docs and mess around with various features (like disabling LSP features). I'm hoping this discussion can save me (an thousands of others) a bunch of time by serving as a reference.

Using emacs-pgtk-next 29.0.50 on GNU Guix System
Eglot version 20220630.739
LSP Server of largest concern: (stupid) typescript language server.

**Comment by trev-dev** [comment, marked as answer]
Okay, here's my "performance tweaked" configuration. It's not terminal friendly (yet) but where I spend 95% of my time, it's working well.

I tested it in the largest, most questionable pre-processed JS file from a CMS website I could find in 2 minutes. It was only 229kb in size. I expect files like these to have performance problems everywhere (VSCode, NeoVIm, LSP-Mode, Eglot). It did stutter, but it wasn't too much for me to feel annoyed. What's important is that normal files that people enjoy working on did not bother me.

The config uses:
1. Eldoc Box
2. Eglot (obviously)
3. Cape/Corfu with _manual_ completion.

Corfu:
```elisp
  (require 'corfu)
  (defun corfu-enable-in-minibuffer ()
    "Enable Corfu in the minibuffer if `completion-at-point' is bound."
    (when (where-is-internal #'completion-at-point (list (current-local-map)))
      (corfu-mode 1)))

  (add-hook 'eshell-mode-hook
            (lambda ()
              (setq-local corfu-auto nil)
              (corfu-mode)))

  (defun corfu-send-shell (&rest _)
    "Send completion candidate when inside comint/eshell."
    (cond
     ((and (derived-mode-p 'eshell-mode) (fboundp 'eshell-send-input))
      (eshell-send-input))
     ((and (derived-mode-p 'comint-mode)  (fboundp 'comint-send-input))
      (comint-send-input))))
  (advice-add #'corfu-insert :after #'corfu-send-shell)

  ;; Silence the pcomplete capf, no errors or messages!
  (advice-add 'pcomplete-completions-at-point :around #'cape-wrap-silent)

  ;; Ensure that pcomplete does not write to the buffer
  ;; and behaves as a pure `completion-at-point-function'.
  (advice-add 'pcomplete-completions-at-point :around #'cape-wrap-purify)

  (add-hook 'minibuffer-setup-hook #'corfu-enable-in-minibuffer)

  (with-eval-after-load 'corfu
    (define-key corfu-map (kbd "SPC") #'corfu-insert-separator))
  (global-corfu-mode)
  ```
  
  Corfu:
  ```elisp
  (with-eval-after-load 'corfu
    (require 'cape)

    (defun td/don-local-cape (comps)
      "Return a hook function to set local capfs."
      `(lambda ()
         (setq-local completion-at-point-functions ',comps)))

    (defvar td/cape-base (list #'cape-file
                               #'cape-history
                               #'cape-dabbrev)
      "Base completions back-ends.")
    (defvar td/cape-writing (append td/cape-base
                                    (list #'cape-ispell))
      "Completions used for writing modes.")
    (defvar td/cape-prog (append td/cape-base
                                 (list #'cape-symbol
                                       #'cape-keyword))
      "Completions for programming modes.")
    (dolist (comp td/cape-base)
      (add-to-list 'completion-at-point-functions comp))
    (with-eval-after-load 'cape
      (add-hook 'prog-mode-hook (td/don-local-cape td/cape-prog))
      (add-hook 'org-mode-hook (td/don-local-cape td/cape-writing))
      (add-hook 'message-mode-hook (td/don-local-cape td/cape-writing))))
```

Eldoc/Box:
```elisp
;; Prevent eldoc from resizing.
(setq eldoc-echo-area-use-multiline-p nil)
(require 'eldoc-box)
;; Only use helper functions to call up information when I want more details.
;; Avoid using automatic pop-up modes like eldoc-box-hover-mode. They
;; hurt buffer performance like nothing else in the LSP space.
(global-set-key (kbd "C-c k") #'eldoc-box-help-at-point)
(define-key eglot-mode-map (kbd "C-c K") #'eldoc-box-eglot-help-at-point)
```

Eglot:
```elisp
(with-eval-after-load 'eglot
  (add-to-list 'eglot-server-programs '(php-mode . ("intelephense" "--stdio")))
  (add-to-list 'eglot-server-programs '(svelte-mode . ("svelteserver" "--stdio")))
  (add-to-list 'eglot-server-programs '(shopify-mode
                                        . ("theme-check-language-server" "--stdio")))
  (td/bind-keys '(("C-c e r" . eglot-rename)
                  ("C-c e d" . eglot-find-typeDefinition)
                  ("C-c e D" . eglot-find-declaration)
                  ("C-c e f" . eglot-format)
                  ("C-c e F" . eglot-format-buffer)
                  ("C-c e R" . eglot-reconnect)) eglot-mode-map)
  
  ;; No event buffers, disable providers cause a lot of hover traffic. Shutdown unused servers.
  (setq eglot-events-buffer-size 0
        eglot-ignored-server-capabilities '(:hoverProvider
                                            :documentHighlightProvider)
        eglot-autoshutdown t))
;; Show all of the available eldoc information when we want it. This way Flymake errors
;; don't just get clobbered by docstrings.        
(add-hook 'eglot-managed-mode-hook
          (lambda ()
            "Make sure Eldoc will show us all of the feedback at point."
            (setq-local eldoc-documentation-strategy
                        #'eldoc-documentation-compose)))
(td/add-hooks '(css-mode
                html-mode
                js-mode
                python-mode
                php-mode
                svelte-mode
                typescript-mode) #'eglot-ensure)
```

**Reply by joaotavora** [reply]
Thanks for answering your own question.  I adjusted some excessive vernacular to less potentially offensive words.

**Reply by trev-dev** [reply]
Appreciate it @joaotavora :) I hope I didn't offend anyone. Apologies!

**Comment by joaotavora** [comment]
Hey @trev-dev , I think your question make sense and I empathize with it, since I had the same problem when I started actually using Eglot in earnest and not just developing it for others some time ago.

However similar my problem was to yours (I'm also using the terminal, and I also thought the performance was sometimes poor), my solution is different.  I use eglot.el + company.el and nothing else (and sometimes I skip company entirely and just use good old C-M-i)

I don't use `eglot-ensure`, because typing `M-x eglot`  once in the morning or every time I open a new project isn't too much work and I rarely shut down my Emacs.

I set these variables which make Eglot do less requests to the server and thus normal operations are more responsive

```
(setq eldoc-idle-delay 0.75)
(setq company-idle-delay 0.75)
(setq flymake-no-changes-timeout 0.5)
```

And that's basically it for performance.  I find it's pretty spiffy, even on large projects.  I have a decent 6yo thinkpad, nothing special. Completions take a while to appear and to update, but I don't mind this, it meshes well with my flow of guessing and confirming in under a second.

To stay sane with the Eldoc dance, I'll frequently set `(setq eldoc-area-prefer-doc-buffer to t)` and keep an eldoc buffer open in some window.  Then the echo area doesn't jump around . But I don't mind if it jumps around when that's all there is to show documentation.

I also use `flymake-goto-next-error` and `flymake-goto-prev-error` bound to `M-n` and `M-p`, a old habit from my SLIME days.

**Reply by trev-dev** [reply]
Thanks for sharing your strategy. I will definitely be giving those variables a try :) I enjoy most often having 2 windows open in a frame, both usually being different code files (or perhaps one in 2 positions), which is why I enjoy things like "eldoc-box".

I have a 7 year old Dell XPS, it's still a decent machine. What keeps you running emacs in -nw?

**Comment by trev-dev** [comment]
In my other (mis)adventures I learned about profiling, which helped me get to the bottom of a separate performance issue.

I circled back to my LSP/javascript woes, and did some profiling. It turns out that the biggest hit to my CPU performance comes from the use of `js2-mode` and it's buffer parsing  :open_mouth: I have turned that off and stuck to the internal javascript-mode  & tree-sitter.  The bigger/bloated JS files have significantly improved in performance.

With or without some features from Eglot/LSP, I am a lot happier now. Thanks @joaotavora again for your suggestions.

**Reply by joaotavora** [reply]
Nice choice, I stay away from `js2-mode` for that reason, too.  Nice, but too complex (and js syntax has evolved a lot).  Is the `tree-sitter` thing already usable in Emacs 29??  I'm very curious.

**Reply by trev-dev** [reply]
Assuming we are [talking about the same thing](https://emacs-tree-sitter.github.io/), I have been using emacs-tree-sitter since around Emacs 26. I do get prettier/more consistent syntax highlighting, but it's not perfect. The imperfections aren't bad enough to make me not want to use it :smile:
